File: packages/python/src/tbd_agents/resources/tasks.py
from __future__ import annotations

from tbd_agents.types import TaskExecution, TaskProgress, TaskExecutionSummary

from .base import BaseResource


class TasksResource(BaseResource):
    def list(self) -> list[TaskExecutionSummary]:
        return self._get_list("/tasks", model=TaskExecutionSummary)

    def get(self, task_id: str) -> TaskExecution:
        return self._get(f"/tasks/{task_id}", model=TaskExecution)

    def progress(self, task_id: str) -> TaskProgress:
        return self._get(f"/tasks/{task_id}/progress", model=TaskProgress)

    def list_for_workflow(self, workflow_id: str) -> list[TaskExecutionSummary]:
        return self._get_list(f"/tasks/workflow/{workflow_id}", model=TaskExecutionSummary)

File: packages/typescript/src/resources/tasks.ts
import { BaseResource, pollUntil } from './base.js';
import type {
  QueryParams,
  TaskExecution,
  TaskProgress,
  TaskSummary,
} from '../types.js';

export interface TaskPollOptions {
  intervalMs?: number;
  timeoutMs?: number;
  signal?: AbortSignal;
}

export class TasksResource extends BaseResource {
  list(query?: QueryParams): Promise<TaskSummary[]> {
    return this.client.request<TaskSummary[]>(
      query
        ? {
            path: 'tasks',
            query,
          }
        : {
            path: 'tasks',
          },
    );
  }

  listForWorkflow(workflowId: string): Promise<TaskSummary[]> {
    return this.client.request<TaskSummary[]>({
      path: `tasks/workflow/${encodeURIComponent(workflowId)}`,
    });
  }

  get(id: string): Promise<TaskExecution> {
    return this.client.request<TaskExecution>({
      path: `tasks/${encodeURIComponent(id)}`,
    });
  }

  progress(id: string): Promise<TaskProgress> {
    return this.client.request<TaskProgress>({
      path: `tasks/${encodeURIComponent(id)}/progress`,
    });
  }

  waitForStatus(
    id: string,
    statuses: string[],
    options?: TaskPollOptions,
  ): Promise<TaskExecution> {
    return pollUntil(() => this.get(id), {
      ...options,
      predicate: (task) => statuses.includes(task.status),
    });
  }

  waitForCompletion(id: string, options?: TaskPollOptions): Promise<TaskExecution> {
    return this.waitForStatus(id, ['completed', 'failed', 'cancelled'], options);
  }
}
